pkg/controllers/vdb/replicatedupgrade_reconciler.go

```diff
@@ -93,42 +101,192 @@ func (r *ReplicatedUpgradeReconciler) Reconcile(ctx context.Context, _ *ctrl.Req
+		if sc.Type != vapi.PrimarySubcluster {
+			continue
+		}
+
+		newsc := sc.DeepCopy()
+		newSCName, err := r.genNewSubclusterName(sc.Name, scMap)
+		if err != nil {
+			return false, err
+		}
+		newsc.Name = newSCName
+		// The subcluster will be sandboxed. And only secondaries can be
+		// sandbox.
+		newsc.Type = vapi.SecondarySubcluster
+		// We don't want to duplicate the service object settings. These new
+		// subclusters will eventually reuse the service object of the primaries
+		// they are mimicing. But not until they are ready to accept
+		// connections. In the meantime, we will setup a simple ClusterIP style
+		// service object. No one should really be connecting to them.
+		newsc.ServiceType = corev1.ServiceTypeClusterIP
+		newsc.ClientNodePort = 0
+		newsc.ExternalIPs = nil
+		newsc.LoadBalancerIP = ""
+		newsc.ServiceAnnotations = nil
+		newsc.ServiceName = ""
+		newsc.VerticaHTTPNodePort = 0
+		// The image in the vdb has already changed to the new one. We need to
+		// set the image override so that the new subclusters come up with the
+		// old image.
+		newsc.ImageOverride = oldImage
 
-	for inx := range stss.Items {
-		sts := &stss.Items[inx]
+		// Include annotations to indicate what replica group it is assigned to
+		// and provide a link back to the subcluster it is defined from.
+		if newsc.Annotations == nil {
+			newsc.Annotations = make(map[string]string)
+		}
+		newsc.Annotations[vmeta.ReplicaGroupAnnotation] = vmeta.ReplicaGroupBValue
+		newsc.Annotations[vmeta.ParentSubclusterAnnotation] = sc.Name
 
-		scName, ok := sts.Labels[vmeta.SubclusterNameLabel]
-		if !ok {
-			return ctrl.Result{},
-				fmt.Errorf("statefulset %q has missing subcluster name label %q", sts.Name, vmeta.SubclusterNameLabel)
+		// Create a linkage in the parent-child
+		if sc.Annotations == nil {
+			sc.Annotations = make(map[string]string)
 		}
+		sc.Annotations[vmeta.ChildSubclusterAnnotation] = newsc.Name
 
-		upgradeStatus.ReplicaGroups[replicaGroupA] = append(upgradeStatus.ReplicaGroups[replicaGroupA], scName)
+		scMap[newSCName] = newsc
+		newSubclusters = append(newSubclusters, *newsc)
 	}
 
-	// Commit the replica groups to the status field for subsequent steps to pick up.
-	err = vdbstatus.SetUpgradeState(ctx, r.VRec.Client, r.VDB, &upgradeStatus)
-	return ctrl.Result{}, err
+	if len(newSubclusters) == 0 {
+		return false, errors.New("no primary subclusters found")
+	}
+	r.VDB.Spec.Subclusters = append(r.VDB.Spec.Subclusters, newSubclusters...)
+	return true, nil
+}
+
+// assignSubclustersToReplicaGroupACallback is a callback method to update the
+// VDB. It will assign each subcluster to replica group A by setting an
+// annotation.
+func (r *ReplicatedUpgradeReconciler) assignSubclustersToReplicaGroupACallback() (bool, error) {
+	annotatedAtLeastOnce := false
+	for inx := range r.VDB.Spec.Subclusters {
+		sc := &r.VDB.Spec.Subclusters[inx]
+		if val, found := sc.Annotations[vmeta.ReplicaGroupAnnotation]; !found ||
```

Comment:
Shouldn't check if the annotations == nil first?

Reply:
No. You can always do lookups into a nil map. You'll never find a key obviously, but its valid Go code to do.